skimage/morphology/selem.py

```diff
@@ -1,366 +1,13 @@
-    n : int
-        The height or width of the slanted sides.
-
-    Other Parameters
-    ----------------
-    dtype : data-type
-        The data type of the structuring element.
-
-    Returns
-    -------
-    selem : ndarray
-        The structuring element where elements of the neighborhood
-        are 1 and 0 otherwise.
-
-    """
-    from . import convex_hull_image
-    selem = np.zeros((m + 2 * n, m + 2 * n))
-    selem[0, n] = 1
-    selem[n, 0] = 1
-    selem[0, m + n - 1] = 1
-    selem[m + n - 1, 0] = 1
-    selem[-1, n] = 1
-    selem[n, -1] = 1
-    selem[-1, m + n - 1] = 1
-    selem[m + n - 1, -1] = 1
-    selem = convex_hull_image(selem).astype(dtype)
-    return selem
-
-
-def star(a, dtype=np.uint8):
-    """Generates a star shaped structuring element.
-
-    Start has 8 vertices and is an overlap of square of size `2*a + 1`
-    with its 45 degree rotated version.
-    The slanted sides are 45 or 135 degrees to the horizontal axis.
-
-    Parameters
-    ----------
-    a : int
-        Parameter deciding the size of the star structural element. The side
-        of the square array returned is `2*a + 1 + 2*floor(a / 2)`.
-
-    Other Parameters
-    ----------------
-    dtype : data-type
-        The data type of the structuring element.
-
-    Returns
-    -------
-    selem : ndarray
-        The structuring element where elements of the neighborhood
-        are 1 and 0 otherwise.
-
-    """
-    from . import convex_hull_image
-
-    if a == 1:
-        bfilter = np.zeros((3, 3), dtype)
-        bfilter[:] = 1
-        return bfilter
-
-    m = 2 * a + 1
-    n = a // 2
-    selem_square = np.zeros((m + 2 * n, m + 2 * n))
-    selem_square[n: m + n, n: m + n] = 1
-
-    c = (m + 2 * n - 1) // 2
-    selem_rotated = np.zeros((m + 2 * n, m + 2 * n))
-    selem_rotated[0, c] = selem_rotated[-1, c] = 1
-    selem_rotated[c, 0] = selem_rotated[c, -1] = 1
-    selem_rotated = convex_hull_image(selem_rotated).astype(int)
-
-    selem = selem_square + selem_rotated
-    selem[selem > 0] = 1
-
-    return selem.astype(dtype)
-
-
-def _default_selem(ndim):
-    """Generates a cross-shaped structuring element (connectivity=1).
-
-    This is the default structuring element (selem) if no selem was specified.
-
-    Parameters
-    ----------
-    ndim : int
-        Number of dimensions of the image.
-
-    Returns
-    -------
-    selem : ndarray
-        The structuring element where elements of the neighborhood
-        are 1 and 0 otherwise.
-
-    """
-    return ndi.generate_binary_structure(ndim, 1)
+import warnings
+
+from .footprint import (
+    square, rectangle, diamond, disk, cube, octahedron, ball, octagon, star
+)
+
+warnings.warn(
+    "The `skimage.morphology.selem` module is deprecated and will be removed "
+    "in scikit-image 1.0. The functions formerly in "
+    "`skimage.morphology.selem` should be imported directly from "
+    "`skimage.morphology` instead.",
+    FutureWarning, stacklevel=2
+)
```

Comment:
Lovely.

I wonder though whether we should not instead encourage directly using `from skimage.morphology import footprints; footprint = footprints.square(5)`. To me this addresses the point in #5439 about grouping functions according to "purpose and equivalence classes". `morphology.square()` is also less descriptive than `footprints.square`.

Perhaps something to add to a dedicated issue for discussion (tracked from #5439) rather than hold up this PR.

Reply:
(But, a quick fix here would be to remain agnostic in the prescription: instead of "should be imported from skimage.morphology instead", we could write "skimage.morphology.selem has been moved to skimage.morphology.footprints.") (And yes I have a mild preference for "footprints" over "footprint", because the latter is easily confused with a variable/kwarg name.)

Reply:
I was not sure if this `selem`/`footprint` module within `filters`, `morphology`, etc. was really intended to be public or not. Specifically, these functions are [grouped in with all other morphology functions in the docs](https://scikit-image.org/docs/stable/api/skimage.morphology.html).

I am fine with renaming to `footprints.py` and importing from there, but wonder if we can perhaps follow up later to make these logical groupings (`binary` vs. `gray` vs. `footprints`) clearer in the docs.

Reply:
Yes, following things up later is totally fine, but that's why I'm saying, let's not be prescriptive in the deprecation message, because we might change our mind during those discussions.